subprojects/file-collections/src/main/java/org/gradle/api/internal/file/DefaultFilePropertyFactory.java

```diff
@@ -348,12 +348,7 @@ public Provider<Directory> dir(final String path) {
 
         @Override
         public Provider<Directory> dir(final Provider<? extends CharSequence> path) {
-            return new AbstractCombiningProvider<Directory, Directory, CharSequence>(Directory.class, this, Providers.internal(path)) {
-                @Override
-                protected Directory map(Directory b, CharSequence v) {
-                    return b.dir(v.toString());
-                }
-            };
+            return new BiProvider<>(Directory.class, this, path, (dir, relativePath) -> dir.dir(relativePath.toString()));
```

Comment:
The only uses of `AbstractCombiningProvider` were in this file, and the functionality of this provider 100% matches the one of `BiProvider`. The only difference was that `BiProvider` never returned any type from `getType()`. This got fixed in the same PR and `AbstractCombiningProvider` became obsolete.